core/primitives-core/src/runtime/fees.rs

```diff
@@ -203,28 +204,51 @@ impl StorageUsageConfig {
 
 /// Helper functions for computing Transfer fees.
 /// In case of implicit account creation they always include extra fees for the CreateAccount and
-/// AddFullAccessKey actions that are implicit.
+/// AddFullAccessKey (for NEAR-implicit account only) actions that are implicit.
 /// We can assume that no overflow will happen here.
-pub fn transfer_exec_fee(cfg: &RuntimeFeesConfig, is_receiver_implicit: bool) -> Gas {
-    if is_receiver_implicit {
-        cfg.fee(ActionCosts::create_account).exec_fee()
-            + cfg.fee(ActionCosts::add_full_access_key).exec_fee()
-            + cfg.fee(ActionCosts::transfer).exec_fee()
-    } else {
-        cfg.fee(ActionCosts::transfer).exec_fee()
+pub fn transfer_exec_fee(
+    cfg: &RuntimeFeesConfig,
+    implicit_account_creation_allowed: bool,
+    receiver_account_type: AccountType,
+) -> Gas {
+    let transfer_fee = cfg.fee(ActionCosts::transfer).exec_fee();
+    let create_account_fee = cfg.fee(ActionCosts::create_account).exec_fee();
+    let add_access_key_fee = cfg.fee(ActionCosts::add_full_access_key).exec_fee();
```

Comment:
nit: how expensive is exec_fee()? it seems we are executing it even when they may not be used depending on the result of match?

Reply:
It seems to be free:
```
pub struct Fee {
    pub send_sir: Gas,
    pub send_not_sir: Gas,
    pub execution: Gas,
}

impl Fee {
    ...
    pub fn exec_fee(&self) -> Gas {
        self.execution
    }
```
Although, it might safer to duplicate `cfg.fee(ActionCosts::create_account).exec_fee()` in case `exec_fee()` becomes expensive in future. I will change it.

Reply:
I think it's totally fine and the improved readability totally worth it.

Reply:
Either is fine for me so feel free to merge when you're ready :)